### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by Larisa-sh**

### CODE REVIEW

File: src/main/java/com/example/springbootapp/controller/EmployeeController.java [hunk 1]
```diff
@@ -26,22 +30,34 @@ public List<EmployeeDTO> getAllEmployees() {
...
         return employeeService.saveNewEmployee(employeeDTO);
     }
 
     @PutMapping("/{employeeId}")
-    public Employee updateEmployee(@PathVariable("employeeId") long employeeId, @RequestBody EmployeeDTO employeeDTO) {
+    public Employee updateEmployee(@PathVariable("employeeId")
+                                   @Min(value = 1,
+                                           message = "Employee ID can't be less than 1")
+                                   long employeeId,
+                                   @Valid
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/controller/EmployeeController.java]: Do we really need it?

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java [hunk 2]
```diff
@@ -33,6 +39,37 @@ public ResponseError employeeNotSaved(EmployeeNotSavedException exception) {
...
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ResponseBody
+    public ValidationResponseError onMethodArgumentNotValidException(
+            MethodArgumentNotValidException exception
+    ) {
+        LOGGER.error(exception.getMessage(), exception);
+        final List<Violation> violations = exception.getBindingResult().getFieldErrors().stream()
+                .map(error -> new Violation(error.getField(), error.getDefaultMessage()))
+                .collect(Collectors.toList());
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: Just toList(). Please check other places

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java [hunk 3]
```diff
@@ -33,6 +39,37 @@ public ResponseError employeeNotSaved(EmployeeNotSavedException exception) {
...
     }
 
+    @ResponseBody
+    @ExceptionHandler(ConstraintViolationException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ValidationResponseError onConstraintValidationException(
+            ConstraintViolationException exception
+    ) {
+        LOGGER.error(exception.getMessage(), exception);
+        final List<Violation> violations = exception.getConstraintViolations().stream()
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: Extract this logic in a separate private method please. The same for the onMethodArgumentNotValidException method.

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java [hunk 4]
```diff
@@ -33,6 +39,37 @@ public ResponseError employeeNotSaved(EmployeeNotSavedException exception) {
         return new ResponseError(HttpStatus.BAD_REQUEST, exception.getMessage());
     }
 
+    @ResponseBody
+    @ExceptionHandler(ConstraintViolationException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ValidationResponseError onConstraintValidationException(
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: Please unify method names

File: src/main/java/com/example/springbootapp/exception/Violation.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.example.springbootapp.exception;
+
+public record Violation(String fieldName, String message) {
+
+    public String getFieldName() {
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/exception/Violation.java]: Why do you need this method?

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java [hunk 6]
```diff
@@ -33,6 +39,37 @@ public ResponseError employeeNotSaved(EmployeeNotSavedException exception) {
...
+        LOGGER.error(exception.getMessage(), exception);
+        final List<Violation> violations = exception.getConstraintViolations().stream()
+                .map(
+                        violation -> new Violation(
+                                violation.getPropertyPath().toString(),
+                                violation.getMessage()
+                        )
+                )
+                .collect(Collectors.toList());
+        return new ValidationResponseError(HttpStatus.BAD_REQUEST,violations);
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: We discussed that we need the same object for all error responses. Why did you decide to use the other object?

**Review by DzianisYermalovich — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/springbootapp/controller/EmployeeController.java (hunk 1 above)
Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/controller/EmployeeController.java]: No, bc there are another code which caught this kind of exeption

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java (hunk 4 above)
Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: done

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java (hunk 2 above)
Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: done

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java (hunk 6 above)
Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: Bc of good list of violations, but i changed it to ResponseError by converting List of violations to String

File: src/main/java/com/example/springbootapp/exception/Violation.java (hunk 5 above)
Comment (DzianisYermalovich) [reply src/main/java/com/example/springbootapp/exception/Violation.java]: The same question for getMessage and toString

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java (hunk 6 above)
Comment (DzianisYermalovich) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: Please add the optional List<Violation> field in ResponseError and use it when you have the list of such violations

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java (hunk 3 above)
Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: I did it.

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java (hunk 6 above)
Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: optional, do you mean java.util.Optional?

File: src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java [hunk 7]
```diff
@@ -33,6 +38,44 @@ public ResponseError employeeNotSaved(EmployeeNotSavedException exception) {
...
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ResponseError employeeConstraintViolation(
+            ConstraintViolationException exception
+    ) {
+        final List<Violation> violations = extractConstraintViolations(exception);
+        return new ResponseError(HttpStatus.BAD_REQUEST, violations);
+    }
+
+    private List<Violation> extractConstraintViolations(ConstraintViolationException exception) {
+        LOGGER.error(exception.getMessage(), exception);
```
Comment (DzianisYermalovich) [inline src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: Please, move logging into the main method(employeeConstraintViolation). The same for the next method

Comment (Larisa-sh) [reply src/main/java/com/example/springbootapp/exception/EmployeeExceptionHandler.java]: I moved it.

**Pull request merged**